<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>GWAS · SequentialGWAS.jl</title><meta name="title" content="GWAS · SequentialGWAS.jl"/><meta property="og:title" content="GWAS · SequentialGWAS.jl"/><meta property="twitter:title" content="GWAS · SequentialGWAS.jl"/><meta name="description" content="Documentation for SequentialGWAS.jl."/><meta property="og:description" content="Documentation for SequentialGWAS.jl."/><meta property="twitter:description" content="Documentation for SequentialGWAS.jl."/><meta property="og:url" content="https://baillielab.github.io/SequentialGWAS.jl/gwas/"/><meta property="twitter:url" content="https://baillielab.github.io/SequentialGWAS.jl/gwas/"/><link rel="canonical" href="https://baillielab.github.io/SequentialGWAS.jl/gwas/"/><script data-outdated-warner src="../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../search_index.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-mocha.css" data-theme-name="catppuccin-mocha"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-macchiato.css" data-theme-name="catppuccin-macchiato"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-frappe.css" data-theme-name="catppuccin-frappe"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-latte.css" data-theme-name="catppuccin-latte"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../">SequentialGWAS.jl</a></span></div><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li><a class="tocitem" href="../">Home</a></li><li><span class="tocitem">Workflows</span><ul><li><a class="tocitem" href="../prerequisites/">Prerequisites</a></li><li><a class="tocitem" href="../combining_datasets/">Combining Datasets</a></li><li><a class="tocitem" href="../genotypes_imputation/">Genotypes Imputation</a></li><li><a class="tocitem" href="../combining_with_ukb/">Combining with UK Biobank</a></li><li class="is-active"><a class="tocitem" href>GWAS</a><ul class="internal"><li><a class="tocitem" href="#Inputs"><span>Inputs</span></a></li><li><a class="tocitem" href="#Workflow-Parameters"><span>Workflow Parameters</span></a></li><li><a class="tocitem" href="#Running-The-Workflow"><span>Running The Workflow</span></a></li><li><a class="tocitem" href="#Outputs"><span>Outputs</span></a></li><li><a class="tocitem" href="#Workflow-DAG"><span>Workflow DAG</span></a></li></ul></li></ul></li><li><span class="tocitem">For Developpers</span><ul><li><a class="tocitem" href="../development/">Development</a></li><li><a class="tocitem" href="../mock_data/">Mock Data</a></li><li><a class="tocitem" href="../julia_fns/">Index Of Julia Functions</a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">Workflows</a></li><li class="is-active"><a href>GWAS</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>GWAS</a></li></ul></nav><div class="docs-right"><a class="docs-navbar-link" href="https://github.com/baillielab/sequential-gwas" title="View the repository on GitHub"><span class="docs-icon fa-brands"></span><span class="docs-label is-hidden-touch">GitHub</span></a><a class="docs-navbar-link" href="https://github.com/baillielab/sequential-gwas/blob/main/docs/src/gwas.md" title="Edit source on GitHub"><span class="docs-icon fa-solid"></span></a><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><h1 id="GWAS"><a class="docs-heading-anchor" href="#GWAS">GWAS</a><a id="GWAS-1"></a><a class="docs-heading-anchor-permalink" href="#GWAS" title="Permalink"></a></h1><h2 id="Inputs"><a class="docs-heading-anchor" href="#Inputs">Inputs</a><a id="Inputs-1"></a><a class="docs-heading-anchor-permalink" href="#Inputs" title="Permalink"></a></h2><h3 id="Data"><a class="docs-heading-anchor" href="#Data">Data</a><a id="Data-1"></a><a class="docs-heading-anchor-permalink" href="#Data" title="Permalink"></a></h3><p>For this workflow, both bed and bgen genotypes should be available. They will likely be the respective outputs of the <a href="../combining_datasets/#Combining-Datasets">Combining Datasets</a> and <a href="../genotypes_imputation/#Genotypes-Imputation">Genotypes Imputation</a> workflows, optionally merged with <a href="../combining_with_ukb/#Combining-with-UK-Biobank">Combining with UK Biobank</a> depending on your project of interest.</p><h3 id="GWAS-Variables"><a class="docs-heading-anchor" href="#GWAS-Variables">GWAS Variables</a><a id="GWAS-Variables-1"></a><a class="docs-heading-anchor-permalink" href="#GWAS-Variables" title="Permalink"></a></h3><p>The variables to be used for the GWAS are specified in a YAML file like so.</p><pre><code class="language-yaml hljs">group: [&quot;ANCESTRY&quot;]
phenotype: &quot;SEVERE_COVID_19&quot;
covariates: [&quot;AGE&quot;, &quot;SEX&quot;, &quot;AGE_x_SEX&quot;, &quot;AGE_x_AGE&quot;]</code></pre><p>where:</p><ul><li>A GWAS will be run independently for each group defined by <code>group</code></li><li>The phenotype is the trait of interest, currently only <code>SEVERE_COVID_19</code></li><li>The covariates are additional covariates that will enter the GWAS to improve precision</li></ul><h2 id="Workflow-Parameters"><a class="docs-heading-anchor" href="#Workflow-Parameters">Workflow Parameters</a><a id="Workflow-Parameters-1"></a><a class="docs-heading-anchor-permalink" href="#Workflow-Parameters" title="Permalink"></a></h2><p>This is the list of all the pipeline&#39;s parameters, they can be set in the <code>run.config</code> file under the <code>params</code> section.</p><ul><li><code>GENOTYPES_PREFIX</code>: Prefix to plink <code>.bed</code> genotypes.</li><li><code>BGEN_GENOTYPES_PREFIX</code> = Prefix to imputed <code>.bgen</code> genotypes.</li><li><code>COVARIATES</code>: Path to covariate file (likely the output of <a href="../combining_datasets/#Combining-Datasets">Combining Datasets</a>)</li><li><code>N_PCS (default: 10)</code>: Number of principal components to compute.</li><li><code>PCA_APPROX</code> (default: true): Whether PCA is performed via approximation <a href="https://www.cog-genomics.org/plink/2.0/strat">see</a></li><li><code>MIN_GROUP_SIZE (default: 100)</code>: Minimum number of samples in a group to proceed to effect size estimation.</li><li><code>VARIABLES_CONFIG (default: assets/variables.yaml)</code>: File containing the declaration of groups, phenotypes and covariates for the GWAS (see <a href="#GWAS-Variables">GWAS Variables</a>).</li><li><code>REGENIE_MAF (default: 0.01)</code>: Minor allele frequency for a variant to enter the GWAS.</li><li><code>REGENIE_MAC (default: 10)</code>: Minor allele count for a variant to enter the GWAS.</li><li><code>REGENIE_BSIZE (default: 1000)</code>: Regenie&#39;s block size (see the <a href="https://rgcgithub.github.io/regenie/">regenie docs</a>)</li></ul><h2 id="Running-The-Workflow"><a class="docs-heading-anchor" href="#Running-The-Workflow">Running The Workflow</a><a id="Running-The-Workflow-1"></a><a class="docs-heading-anchor-permalink" href="#Running-The-Workflow" title="Permalink"></a></h2><p>If the previous steps have been completed successfully you can run:</p><pre><code class="language-bash hljs">./run.sh GWAS</code></pre><h2 id="Outputs"><a class="docs-heading-anchor" href="#Outputs">Outputs</a><a id="Outputs-1"></a><a class="docs-heading-anchor-permalink" href="#Outputs" title="Permalink"></a></h2><p>All outputs are produced in <code>PUBLISH_DIR</code>, the main outputs of the workflow are:</p><h2 id="Workflow-DAG"><a class="docs-heading-anchor" href="#Workflow-DAG">Workflow DAG</a><a id="Workflow-DAG-1"></a><a class="docs-heading-anchor-permalink" href="#Workflow-DAG" title="Permalink"></a></h2><iframe src="../assets/gwas_dag.html" width="100%" height="800px"></iframe></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../combining_with_ukb/">« Combining with UK Biobank</a><a class="docs-footer-nextpage" href="../development/">Development »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="auto">Automatic (OS)</option><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option><option value="catppuccin-latte">catppuccin-latte</option><option value="catppuccin-frappe">catppuccin-frappe</option><option value="catppuccin-macchiato">catppuccin-macchiato</option><option value="catppuccin-mocha">catppuccin-mocha</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.10.1 on <span class="colophon-date" title="Friday 4 April 2025 13:43">Friday 4 April 2025</span>. Using Julia version 1.11.4.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
