<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>GWAS · SequentialGWAS.jl</title><meta name="title" content="GWAS · SequentialGWAS.jl"/><meta property="og:title" content="GWAS · SequentialGWAS.jl"/><meta property="twitter:title" content="GWAS · SequentialGWAS.jl"/><meta name="description" content="Documentation for SequentialGWAS.jl."/><meta property="og:description" content="Documentation for SequentialGWAS.jl."/><meta property="twitter:description" content="Documentation for SequentialGWAS.jl."/><meta property="og:url" content="https://baillielab.github.io/SequentialGWAS.jl/gwas/"/><meta property="twitter:url" content="https://baillielab.github.io/SequentialGWAS.jl/gwas/"/><link rel="canonical" href="https://baillielab.github.io/SequentialGWAS.jl/gwas/"/><script data-outdated-warner src="../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../search_index.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-mocha.css" data-theme-name="catppuccin-mocha"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-macchiato.css" data-theme-name="catppuccin-macchiato"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-frappe.css" data-theme-name="catppuccin-frappe"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-latte.css" data-theme-name="catppuccin-latte"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../">SequentialGWAS.jl</a></span></div><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li><a class="tocitem" href="../">Home</a></li><li><span class="tocitem">Prereuisistes</span><ul><li><a class="tocitem" href="../odap_prerequisites/">ODAP</a></li><li><a class="tocitem" href="../rap_prerequisites/">UKB RAP</a></li></ul></li><li><span class="tocitem">Workflows</span><ul><li><a class="tocitem" href="../combining_datasets/">Combining Datasets</a></li><li><a class="tocitem" href="../genotypes_imputation/">Genotypes Imputation</a></li><li><a class="tocitem" href="../ukb_merge/">Merging the GenOMICC and UK Biobank Cohorts</a></li><li class="is-active"><a class="tocitem" href>GWAS</a><ul class="internal"><li><a class="tocitem" href="#Inputs"><span>Inputs</span></a></li><li><a class="tocitem" href="#Workflow-Parameters"><span>Workflow Parameters</span></a></li><li><a class="tocitem" href="#Running-The-Workflow"><span>Running The Workflow</span></a></li><li><a class="tocitem" href="#Outputs"><span>Outputs</span></a></li><li><a class="tocitem" href="#Workflow-DAG"><span>Workflow DAG</span></a></li></ul></li></ul></li><li><span class="tocitem">For Developpers</span><ul><li><a class="tocitem" href="../development/">Development</a></li><li><a class="tocitem" href="../mock_data/">Mock Data</a></li><li><a class="tocitem" href="../julia_fns/">Index Of Julia Functions</a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">Workflows</a></li><li class="is-active"><a href>GWAS</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>GWAS</a></li></ul></nav><div class="docs-right"><a class="docs-navbar-link" href="https://github.com/baillielab/sequential-gwas" title="View the repository on GitHub"><span class="docs-icon fa-brands"></span><span class="docs-label is-hidden-touch">GitHub</span></a><a class="docs-navbar-link" href="https://github.com/baillielab/sequential-gwas/blob/main/docs/src/gwas.md" title="Edit source on GitHub"><span class="docs-icon fa-solid"></span></a><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><h1 id="GWAS"><a class="docs-heading-anchor" href="#GWAS">GWAS</a><a id="GWAS-1"></a><a class="docs-heading-anchor-permalink" href="#GWAS" title="Permalink"></a></h1><h2 id="Inputs"><a class="docs-heading-anchor" href="#Inputs">Inputs</a><a id="Inputs-1"></a><a class="docs-heading-anchor-permalink" href="#Inputs" title="Permalink"></a></h2><p>This section describes the main inputs to the workflow, see further down for the associated Nextflow parameters.</p><h3 id="Genetic-Data"><a class="docs-heading-anchor" href="#Genetic-Data">Genetic Data</a><a id="Genetic-Data-1"></a><a class="docs-heading-anchor-permalink" href="#Genetic-Data" title="Permalink"></a></h3><p>For this workflow, both genotypes and imputed genotypes are required. They will likely be the respective outputs of the <a href="../combining_datasets/#Combining-Datasets">Combining Datasets</a> and <a href="../genotypes_imputation/#Genotypes-Imputation">Genotypes Imputation</a> workflows, optionally merged with <a href="../combining_with_ukb/#Combining-with-UK-Biobank">Combining with UK Biobank</a> depending on your project of interest.</p><p>### Covariates</p><p>There are two main types of covariates files you can use here. </p><ul><li>The <code>COVARIATES</code> file contains the usual covariates information including phenotypes and is typically provided by Dominique on a per project basis. Unfortunately, at this point, the specifications for this file have not been clearly defined. The columns of the file I have received are: <code>genotype_file_id</code>, <code>age_years</code>, <code>sex</code>, <code>case_or_control</code>, <code>cohort</code>, <code>severe_cohort_primary_diagnosis</code>, <code>isaric_cohort_max_severity_score</code>.</li><li>An optional <code>INFERRED_COVARIATES</code> contains covariates information that is inferred from the genetic datasets during the <a href="../combining_datasets/#Combining-Datasets">Combining Datasets</a> workflow. This is the case for ancestry estimates for instance.</li></ul><h3 id="GWAS-Variables"><a class="docs-heading-anchor" href="#GWAS-Variables">GWAS Variables</a><a id="GWAS-Variables-1"></a><a class="docs-heading-anchor-permalink" href="#GWAS-Variables" title="Permalink"></a></h3><p>You can define the phenotypes, covariates and groups via an external YAML file which looks like the following. </p><pre><code class="language-yaml hljs">groupby: [&quot;ANCESTRY_ESTIMATE&quot;]
phenotypes: [&quot;SEVERE_COVID_19&quot;]
covariates: [&quot;AGE&quot;, &quot;SEX&quot;, &quot;AGE_x_SEX&quot;, &quot;AGE_x_AGE&quot;]</code></pre><p>In this example:</p><ul><li>An independent GWAS will be run for each ancestry group in the <code>ANCESTRY_ESTIMATE</code> column. </li><li>There is only one phenotype: <code>SEVERE_COVID_19</code>. </li><li>The covariates to are <span>$AGE$</span>, <span>$SEX$</span>, <span>$AGE^2$</span> and <span>$AGE \cdot AGE$</span>.</li></ul><h4 id="Groups"><a class="docs-heading-anchor" href="#Groups">Groups</a><a id="Groups-1"></a><a class="docs-heading-anchor-permalink" href="#Groups" title="Permalink"></a></h4><p>The <code>groupby</code> variables defines groups for which an independent GWAS will be run. If the <span>$groupby$</span> key is not present in the YAML file, only one GWAS will be run for the whole population.</p><h4 id="Phenotypes"><a class="docs-heading-anchor" href="#Phenotypes">Phenotypes</a><a id="Phenotypes-1"></a><a class="docs-heading-anchor-permalink" href="#Phenotypes" title="Permalink"></a></h4><p>The <code>phenotype</code> variables must either be <code>case_or_control</code> or have an existing definition based on other columns in the <code>COVARIATES</code> file (hardcoded), at the moment only <code>SEVERE_COVID_19</code> has been defined.</p><h4 id="Covariates"><a class="docs-heading-anchor" href="#Covariates">Covariates</a><a id="Covariates-1"></a><a class="docs-heading-anchor-permalink" href="#Covariates" title="Permalink"></a></h4><p>The <code>covariates</code> variables are used as adjustment variables in the regression model. In REGENIE step 2, principal components are also added in a leave-one-chromosome-out scheme to the covariates defined here. </p><p>Again, due to the ill-definition of the covariates file, at the moment, only a subset of non-inferred covariates can be processed: <code>AGE</code> and <code>SEX</code>. Categorical covariates are one-hot encoded.</p><div class="admonition is-info" id="Interaction-Terms-b45e542cea2d4405"><header class="admonition-header">Interaction Terms<a class="admonition-anchor" href="#Interaction-Terms-b45e542cea2d4405" title="Permalink"></a></header><div class="admonition-body"><p>Interaction terms can be added to the covariate list with the <span>$_x_$</span> delimiter.</p></div></div><h2 id="Workflow-Parameters"><a class="docs-heading-anchor" href="#Workflow-Parameters">Workflow Parameters</a><a id="Workflow-Parameters-1"></a><a class="docs-heading-anchor-permalink" href="#Workflow-Parameters" title="Permalink"></a></h2><p>This is the list of all the pipeline&#39;s parameters, they can be set in the <code>run.config</code> file under the <code>params</code> section.</p><ul><li><code>GENOTYPES_PREFIX</code>: Prefix to genotypes in PLINK BED format (likely the output of the <a href="../combining_datasets/#Combining-Datasets">Combining Datasets</a> workflow).</li><li><code>IMPUTED_GENOTYPES_PREFIX</code>: Prefix to imputed genotypes in PGEN format (likely the output of the <a href="../genotypes_imputation/#Genotypes-Imputation">Genotypes Imputation</a> workflow).</li><li><code>COVARIATES</code>: Path to covariate file (likely the output of the <a href="../combining_datasets/#Combining-Datasets">Combining Datasets</a> workflow)</li><li><code>INFERRED_COVARIATES</code>: Optional, path to covariates inferred from genetic data during the <a href="../combining_datasets/#Combining-Datasets">Combining Datasets</a> workflow.</li><li><code>N_PCS (default: 10)</code>: Number of principal components to compute.</li><li><code>PCA_APPROX</code> (default: true): Whether PCA is performed via approximation <a href="https://www.cog-genomics.org/plink/2.0/strat">see</a></li><li><code>MIN_GROUP_SIZE (default: 100)</code>: Minimum number of samples in a group to proceed to effect size estimation.</li><li><code>VARIABLES_CONFIG (default: assets/variables.yaml)</code>: File containing the declaration of groups, phenotypes and covariates for the GWAS (see <a href="#GWAS-Variables">GWAS Variables</a>).</li><li><code>REGENIE_MAF (default: 0.01)</code>: Minor allele frequency for a variant to enter the GWAS.</li><li><code>REGENIE_MAC (default: 10)</code>: Minor allele count for a variant to enter the GWAS.</li><li><code>REGENIE_BSIZE (default: 1000)</code>: Regenie&#39;s block size (see the <a href="https://rgcgithub.github.io/regenie/">regenie docs</a>)</li></ul><h2 id="Running-The-Workflow"><a class="docs-heading-anchor" href="#Running-The-Workflow">Running The Workflow</a><a id="Running-The-Workflow-1"></a><a class="docs-heading-anchor-permalink" href="#Running-The-Workflow" title="Permalink"></a></h2><p>If the previous steps have been completed successfully you can run on your platform (e.g. <code>PLATFORM=eddie</code>):</p><pre><code class="language-bash hljs">nextflow run main.nf -entry GWAS -resume -profile PLATFORM -with-trace -with-report -c run.config</code></pre><h2 id="Outputs"><a class="docs-heading-anchor" href="#Outputs">Outputs</a><a id="Outputs-1"></a><a class="docs-heading-anchor-permalink" href="#Outputs" title="Permalink"></a></h2><p>All outputs are produced in <code>PUBLISH_DIR (default: results)</code>, for each group a separate folder contains:</p><ul><li>plots: Manhattan and QQ plots</li><li>results: A CSV file of summary statistics</li></ul><h2 id="Workflow-DAG"><a class="docs-heading-anchor" href="#Workflow-DAG">Workflow DAG</a><a id="Workflow-DAG-1"></a><a class="docs-heading-anchor-permalink" href="#Workflow-DAG" title="Permalink"></a></h2><iframe src="../assets/gwas_dag.html" width="100%" height="800px"></iframe></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../ukb_merge/">« Merging the GenOMICC and UK Biobank Cohorts</a><a class="docs-footer-nextpage" href="../development/">Development »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="auto">Automatic (OS)</option><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option><option value="catppuccin-latte">catppuccin-latte</option><option value="catppuccin-frappe">catppuccin-frappe</option><option value="catppuccin-macchiato">catppuccin-macchiato</option><option value="catppuccin-mocha">catppuccin-mocha</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.13.0 on <span class="colophon-date" title="Monday 30 June 2025 10:06">Monday 30 June 2025</span>. Using Julia version 1.11.5.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
