var documenterSearchIndex = {"docs":
[{"location":"odap_prerequisites/#ODAP","page":"ODAP","title":"ODAP","text":"","category":"section"},{"location":"odap_prerequisites/","page":"ODAP","title":"ODAP","text":"The original GenOMICC data is kept on the ODAP platform. To get access, you will need to contact the relevant person, at the moment dominique.mccormick@ed.ac.uk. More information on accessing ODAP can be found here.","category":"page"},{"location":"odap_prerequisites/#Software","page":"ODAP","title":"Software","text":"","category":"section"},{"location":"odap_prerequisites/","page":"ODAP","title":"ODAP","text":"In order to run the workflows in this repository only 2 software need to be installed, and they should already be present on ODAP:","category":"page"},{"location":"odap_prerequisites/","page":"ODAP","title":"ODAP","text":"Nextflow (see how to setup)\nSingularity (Should be ready to use)","category":"page"},{"location":"odap_prerequisites/#Importing-Into-ODAP","page":"ODAP","title":"Importing Into ODAP","text":"","category":"section"},{"location":"odap_prerequisites/","page":"ODAP","title":"ODAP","text":"In order to upload the current state of the code, or a specific version of genomicc-workflows into ODAP, we currently proceed via a shared folder with Dominique.","category":"page"},{"location":"odap_prerequisites/#Locally","page":"ODAP","title":"Locally","text":"","category":"section"},{"location":"odap_prerequisites/","page":"ODAP","title":"ODAP","text":"The following assumes a specific git tag corresponding to a release for which a matching docker image exists, but the steps can be adapted to any need.","category":"page"},{"location":"odap_prerequisites/","page":"ODAP","title":"ODAP","text":"export genomicc_workflows_tag=\"0.1.0\"","category":"page"},{"location":"odap_prerequisites/","page":"ODAP","title":"ODAP","text":"In the shared folder.","category":"page"},{"location":"odap_prerequisites/","page":"ODAP","title":"ODAP","text":"Clone the repository or the relevant commit or tag:","category":"page"},{"location":"odap_prerequisites/","page":"ODAP","title":"ODAP","text":"git clone git@github.com:baillielab/genomicc-workflows.git\ncd genomicc-workflows\ngit checkout v$genomicc_workflows_tag","category":"page"},{"location":"odap_prerequisites/","page":"ODAP","title":"ODAP","text":"Download and save the docker image","category":"page"},{"location":"odap_prerequisites/","page":"ODAP","title":"ODAP","text":"docker pull --platform linux/amd64 olivierlabayle/genomicc:$genomicc_workflows_tag\ndocker save olivierlabayle/genomicc:$genomicc_workflows_tag | gzip > genomicc.tar.gz","category":"page"},{"location":"odap_prerequisites/","page":"ODAP","title":"ODAP","text":"Then ask Dominique to uplaod the folder to ODAP.","category":"page"},{"location":"odap_prerequisites/#On-ODAP","page":"ODAP","title":"On ODAP","text":"","category":"section"},{"location":"odap_prerequisites/","page":"ODAP","title":"ODAP","text":"In the uploaded repository, build the singularity image:","category":"page"},{"location":"odap_prerequisites/","page":"ODAP","title":"ODAP","text":"gunzip genomicc.tar.gz && singularity build genomicc.sif docker-archive:genomicc.tar","category":"page"},{"location":"misc/#Legacy-Information","page":"Legacy","title":"Legacy Information","text":"","category":"section"},{"location":"misc/","page":"Legacy","title":"Legacy","text":"info: Info\nThe data generated by the following guide is not currently used in the aggregation workflow.","category":"page"},{"location":"misc/","page":"Legacy","title":"Legacy","text":"The Illumina manifest files corresponding to each array can be downloaded from Illumina's website. The description of the manifest columns can be found here. A comparison of both the GSA-48v4 and GSA-24v3 arrays for the GRC38 genome build yields the following table (function GenomiccWorkflows.OneTimeChecks.array_overlap): ","category":"page"},{"location":"misc/","page":"Legacy","title":"Legacy","text":"Description Number of SNPs\nIn GSA-48v4 650 321\nIn GSA-24v3 654 027\nUnion 702 515\nIntersection 601 833\nOnly in GSA-48v4 48 488\nOnly in GSA-24v3 52 194","category":"page"},{"location":"misc/","page":"Legacy","title":"Legacy","text":"The arrays are thus quite similar with around 7% difference between the two and none being a subset of the other. This motivates the strategy to take the intersection of genotyping arrays before merging. Note that in reality the intersection might be smaller due to QC filtering of the input SNPs.","category":"page"},{"location":"misc/","page":"Legacy","title":"Legacy","text":"Furthermore, because the manifest files are too big to be version controlled, variants on the - RefStrand were extracted and stored as follows:","category":"page"},{"location":"misc/","page":"Legacy","title":"Legacy","text":"GSA-MD-24v3-0A1 (genome build GRC37) : `assets/GSA-24v3-0A1-minus-strand.txt`\nGSA-MD-48v4-0A1 (genome build GRC38) : `assets/GSA-48v4-020085471_D2-minus-strand.txt`","category":"page"},{"location":"misc/","page":"Legacy","title":"Legacy","text":"This has been done using the make_snps_to_flip_list in the bin/one_time_checks.jl script.","category":"page"},{"location":"mock_data/#Mock-Data","page":"Mock Data","title":"Mock Data","text":"","category":"section"},{"location":"mock_data/","page":"Mock Data","title":"Mock Data","text":"In order to ease testing and development, we generate mock data that closely ressembles the original data while preserving the privacy of individuals in the cohorts. This page explains how this is done.","category":"page"},{"location":"mock_data/","page":"Mock Data","title":"Mock Data","text":"First but foremost, for all individuals, a mock sample ID is created. Since there is no sensitive information about individuals, these cannot be identified as long as the genetic data is perturbed. We explain below how this is done.","category":"page"},{"location":"mock_data/#Genotyping-Arrays","page":"Mock Data","title":"Genotyping Arrays","text":"","category":"section"},{"location":"mock_data/","page":"Mock Data","title":"Mock Data","text":"Only a subset of genetic variations are kept (e.g. 100 out of 500 000)\nFor each individual, each variant is resampled independently from the cohort's empirical distribution. The probability of this operation to have no effect on an individual is difficult to estimate since it depends on each variant's alleles frequencies. If all variants were different it would be (1/n_samples)^nvariants which for the lower values of nsamples=1000 and nvariants=110, this is lower than 10^-300.","category":"page"},{"location":"mock_data/#Whole-Genome-Sequencing","page":"Mock Data","title":"Whole Genome Sequencing","text":"","category":"section"},{"location":"mock_data/","page":"Mock Data","title":"Mock Data","text":"The GVCF mock data arising from whole genome sequencing is built from a very small intersection of variants (e.g., 100) common to all genotyping arrays. Individuals are thus unidentifiable.","category":"page"},{"location":"mock_data/#Covariates","page":"Mock Data","title":"Covariates","text":"","category":"section"},{"location":"mock_data/","page":"Mock Data","title":"Mock Data","text":"Since the covariates are not sensitive, the newly created odap identifier is simply forwarded to covariates.","category":"page"},{"location":"mock_data/#How-to-Mock","page":"Mock Data","title":"How to Mock","text":"","category":"section"},{"location":"mock_data/","page":"Mock Data","title":"Mock Data","text":"The mocked data was generated on ODAP using the mock_data function in the bin/mock.jl script.","category":"page"},{"location":"mock_data/","page":"Mock Data","title":"Mock Data","text":"I also manually:","category":"page"},{"location":"mock_data/","page":"Mock Data","title":"Mock Data","text":"Added duplicate sample IDs to reproduce what is in the data.\nChanged the position of GSA-rs114361133 in test/assets/genomicc/genotyping_arrays/mock.release_2021_2023.map to be unliftable.","category":"page"},{"location":"mock_data/#Thousands-Genomes","page":"Mock Data","title":"Thousands Genomes","text":"","category":"section"},{"location":"mock_data/","page":"Mock Data","title":"Mock Data","text":"The data was downloaded from the 1000GP FTP and pruned using the bin/make_thousand_genomes_filter_files.jl script.","category":"page"},{"location":"development/#Development","page":"Development","title":"Development","text":"","category":"section"},{"location":"development/","page":"Development","title":"Development","text":"This page contains some tips and tricks to help development. Because of the difficulty associated with development on ODAP it is assumed that most of the development will take place outside of it. For that purpose the original data has been mocked and is stored in test/assets. More information on the process can be found in Mock Data","category":"page"},{"location":"development/#Build-the-Docker-Image-Locally","page":"Development","title":"Build the Docker Image Locally","text":"","category":"section"},{"location":"development/","page":"Development","title":"Development","text":"The image is built automatically during the continuous integration process and published on Docker HUB.","category":"page"},{"location":"development/","page":"Development","title":"Development","text":"You can also build it locally if you have docker installed by running the following:","category":"page"},{"location":"development/","page":"Development","title":"Development","text":"docker build -t genomicc-workflows -f docker/Dockerfile .","category":"page"},{"location":"development/","page":"Development","title":"Development","text":"(If running on MacOS with an ARM platform, add: --platform linux/amd64)","category":"page"},{"location":"development/#Code-Development-on-ODAP2","page":"Development","title":"Code Development on ODAP2","text":"","category":"section"},{"location":"development/","page":"Development","title":"Development","text":"When you really need to.","category":"page"},{"location":"development/","page":"Development","title":"Development","text":"First import the docker image as a singularity container within ODAP which we assume is called genomicc-workflows.sif.","category":"page"},{"location":"development/#Julia-REPL","page":"Development","title":"Julia REPL","text":"","category":"section"},{"location":"development/","page":"Development","title":"Development","text":"To get a shell while mounting the repo within the container:","category":"page"},{"location":"development/","page":"Development","title":"Development","text":"singularity shell --bind $PWD:/mnt/genomicc-workflows genomicc-workflows.sif","category":"page"},{"location":"development/","page":"Development","title":"Development","text":"singularity shell --bind $PWD:/mnt/genomicc-workflows genomicc-workflows.sif ","category":"page"},{"location":"development/","page":"Development","title":"Development","text":"Then run the julia REPL","category":"page"},{"location":"development/","page":"Development","title":"Development","text":"JULIA_DEPOT_PATH=$JULIA_DEPOT_PATH:/root/.julia julia --project=/mnt/genomicc-workflows","category":"page"},{"location":"development/#Extra-Tools","page":"Development","title":"Extra Tools","text":"","category":"section"},{"location":"development/","page":"Development","title":"Development","text":"Most tools are available within their conda environment, for instance regenie:","category":"page"},{"location":"development/","page":"Development","title":"Development","text":"docker run -it --rm genomicc-workflows /opt/miniforge3/bin/conda run -n regenie_env regenie --help","category":"page"},{"location":"walk_through/#Walk-Through","page":"Walk Through","title":"Walk Through","text":"","category":"section"},{"location":"walk_through/","page":"Walk Through","title":"Walk Through","text":"In this section we discuss how to regenerate results from an analysis from scratch. Each step consists in a separate workflow for which a dedicated documentation is referenced. While workflows are platform agnostic, due to complex data permissions and governance, each workflow is typically run on a dedicated platform.","category":"page"},{"location":"walk_through/#Step-1:-Aggregating-Genetic-Datasets","page":"Walk Through","title":"Step 1: Aggregating Genetic Datasets","text":"","category":"section"},{"location":"walk_through/","page":"Walk Through","title":"Walk Through","text":"workflow: Combining GenOMICC Datasets\nplatform: ODAP","category":"page"},{"location":"walk_through/#Step-2:-Imputation-of-Genotypes","page":"Walk Through","title":"Step 2: Imputation of Genotypes","text":"","category":"section"},{"location":"walk_through/","page":"Walk Through","title":"Walk Through","text":"workflow: TOPMed Imputation\nplatform: Eddie","category":"page"},{"location":"walk_through/#Step-3:-Merging-GenOMICC-and-the-UK-Biobank","page":"Walk Through","title":"Step 3: Merging GenOMICC and the UK Biobank","text":"","category":"section"},{"location":"walk_through/","page":"Walk Through","title":"Walk Through","text":"workflow: Merging the GenOMICC and UK Biobank Cohorts\nplatform: UKB RAP","category":"page"},{"location":"walk_through/#Step-4:-GWAS","page":"Walk Through","title":"Step 4: GWAS","text":"","category":"section"},{"location":"walk_through/","page":"Walk Through","title":"Walk Through","text":"workflow: WDL-GWAS\nplatform: UKB RAP","category":"page"},{"location":"combining_genomicc_datasets/#Combining-GenOMICC-Datasets","page":"Workflow","title":"Combining GenOMICC Datasets","text":"","category":"section"},{"location":"combining_genomicc_datasets/","page":"Workflow","title":"Workflow","text":"This workflow combines the various genetic data sources available into a unified dataset.","category":"page"},{"location":"combining_genomicc_datasets/#Inputs","page":"Workflow","title":"Inputs","text":"","category":"section"},{"location":"combining_genomicc_datasets/","page":"Workflow","title":"Workflow","text":"Since GenOMICC is an ongoing project where the data is continuously collected, it came and will continue to arrive in different formats. This section describes the input data, the corresponding workflow parameters are described below.","category":"page"},{"location":"combining_genomicc_datasets/#Genetic-Data","page":"Workflow","title":"Genetic Data","text":"","category":"section"},{"location":"combining_genomicc_datasets/","page":"Workflow","title":"Workflow","text":"The pipeline requires genotyping arrays and optionally whole-genome sequencing (wgs) data.","category":"page"},{"location":"combining_genomicc_datasets/#Genotyping-Arrays","page":"Workflow","title":"Genotyping Arrays","text":"","category":"section"},{"location":"combining_genomicc_datasets/","page":"Workflow","title":"Workflow","text":"There are three filesets and three corresponding subfolders:","category":"page"},{"location":"combining_genomicc_datasets/","page":"Workflow","title":"Workflow","text":"The r8 release: Corresponds to genotyping data generated before 2021. The genotyping chip was the Illumina GSA-MD-24v3-0A1 and the associated genome build GRCh37. The corresponding subfolder is usually named `wp5-gwas-r8-under90excl2021Sep16`.\nThe 2021-2023 release:  Corresponds to genotyping data generated between 2021 and 2023. It was also genotyped using the Illumina GSA-MD-24v3-0A1 chip and the genome build is also GRCh37. The corresponding subfolder is usually named `2021092020231206QCVFinal`.\nThe 2024-now release: Corresponds to the latest fileset. The Illumina GSA-MD-48v4-0A1 chip was used and the genome build is GRCh38. The corresponding subfolder is usually named `2024060420240610QCVFinal`.","category":"page"},{"location":"combining_genomicc_datasets/","page":"Workflow","title":"Workflow","text":"In each subfolder, there is a plink subfolder which contains the actual genotypes, the following tables summarises the above","category":"page"},{"location":"combining_genomicc_datasets/","page":"Workflow","title":"Workflow","text":"Brief Description Period Begin Period End Genotyping Array Genome Build Directory Genotypes Prefix\nPrehistoric r8 release 04/05/2020 30/08/2021 GSA-MD-24v3-0_A1 GRC37 wp5-gwas-r8-under90excl_2021Sep16 PLINK1909210906/wp5-gwas-r8-under90excl_2021Sep16\nBefore 2024 microarray 20/09/2021 06/12/2023 GSA-MD-24v3-0_A1 GRC37 2021092020231206QC_VFinal PLINK0407240954/2021092020231206QC_VFinal\nSince 2024 microarray 04/06/2024 10/06/2024 GSA-MD-48v4-0_A1 GRC38 2024060420240610QC_VFinal PLINK0407240114/2024060420240610QC_VFinal","category":"page"},{"location":"combining_genomicc_datasets/","page":"Workflow","title":"Workflow","text":"So, in the example above, the R8_GENOTYPES workflow parameter (see below) should point to wp5-gwas-r8-under90excl_2021Sep16/PLINK_190921_0906/wp5-gwas-r8-under90excl_2021Sep16","category":"page"},{"location":"combining_genomicc_datasets/#Whole-Genome-Sequencing","page":"Workflow","title":"Whole Genome Sequencing","text":"","category":"section"},{"location":"combining_genomicc_datasets/","page":"Workflow","title":"Workflow","text":"The wgs GVCF files are all located in a wgs-reheadered folder.","category":"page"},{"location":"combining_genomicc_datasets/#External-Resources","page":"Workflow","title":"External Resources","text":"","category":"section"},{"location":"combining_genomicc_datasets/","page":"Workflow","title":"Workflow","text":"As well as our in-house data, the pipeline depends on external reference data. In principle these files should already be present on ODAP and there is nothing you need to do.","category":"page"},{"location":"combining_genomicc_datasets/#The-1000-GP","page":"Workflow","title":"The 1000 GP","text":"","category":"section"},{"location":"combining_genomicc_datasets/","page":"Workflow","title":"Workflow","text":"All VCF files and indexes present in this FTP folder\nThe associated 1000 GP pedigree file.","category":"page"},{"location":"combining_genomicc_datasets/","page":"Workflow","title":"Workflow","text":"These files should be stored in the same folder which is defined by the KGP_DIR (default: /mnt/odap-beegfs/software/gwas-resources/1000-genomes-HC) Nextflow parameter.","category":"page"},{"location":"combining_genomicc_datasets/#GATK","page":"Workflow","title":"GATK","text":"","category":"section"},{"location":"combining_genomicc_datasets/","page":"Workflow","title":"Workflow","text":"The reference genome published by the Broad Institute.","category":"page"},{"location":"combining_genomicc_datasets/","page":"Workflow","title":"Workflow","text":"This should be in a folder defined by the GATK_DIR (default: /mnt/odap-beegfs/software/gwas-resources/gatk) Nextflow parameter.","category":"page"},{"location":"combining_genomicc_datasets/#Running-The-Workflow","page":"Workflow","title":"Running The Workflow","text":"","category":"section"},{"location":"combining_genomicc_datasets/","page":"Workflow","title":"Workflow","text":"If the previous steps have been completed successfully you can run:","category":"page"},{"location":"combining_genomicc_datasets/","page":"Workflow","title":"Workflow","text":"taskset -c 998 nextflow run main.nf -entry CombineGeneticDatasets -c run.config -profile odap -resume -with-report -with-trace","category":"page"},{"location":"combining_genomicc_datasets/#Outputs","page":"Workflow","title":"Outputs","text":"","category":"section"},{"location":"combining_genomicc_datasets/","page":"Workflow","title":"Workflow","text":"All outputs are produced in PUBLISH_DIR (defaults to results), the main outputs of the workflow are:","category":"page"},{"location":"combining_genomicc_datasets/","page":"Workflow","title":"Workflow","text":"report.md: A report of the pipeline execution\ngenotypes.aggregated.qced.final.{bed,bim,fam}: The aggregated genotypes.\ncovariates.inferred.csv: The covariates inferred from the genotypes (ancestry, PCs).\nancestry/kgp_shared/kgp.merged.unrelated.shared.{bed,bim,fam}: The 1000 Genome Project genotypes corresponding to the variants in the GenOMICC cohort (can be used later on in Merging the GenOMICC and UK Biobank Cohorts)","category":"page"},{"location":"combining_genomicc_datasets/#Workflow-Parameters","page":"Workflow","title":"Workflow Parameters","text":"","category":"section"},{"location":"combining_genomicc_datasets/","page":"Workflow","title":"Workflow","text":"This is the list of all the pipeline's parameters, they can be set in the run.config file under the params section.","category":"page"},{"location":"combining_genomicc_datasets/#Input-Files","page":"Workflow","title":"Input Files","text":"","category":"section"},{"location":"combining_genomicc_datasets/","page":"Workflow","title":"Workflow","text":"These are project specific and need to be provided:","category":"page"},{"location":"combining_genomicc_datasets/","page":"Workflow","title":"Workflow","text":"R8_GENOTYPES: Prefix to release r8 genotypes (before 2021).\nBEFORE_2024_GENOTYPES: Prefix to genotypes released between 2021-2023.\nSINCE_2024_GENOTYPES: Prefix to genotypes released after 2024.\nWGS_GVCFS (optional): Prefix to whole genome sequencing files.","category":"page"},{"location":"combining_genomicc_datasets/#External-Inputs-Parameters","page":"Workflow","title":"External Inputs Parameters","text":"","category":"section"},{"location":"combining_genomicc_datasets/","page":"Workflow","title":"Workflow","text":"These are already set if you are using the odap profile.","category":"page"},{"location":"combining_genomicc_datasets/","page":"Workflow","title":"Workflow","text":"RESOURCES_DIR (default: ./assets/resources\"): Path to all external resources.\nKGP_DIR: Path to the 1000 Genome Project specific resources (see The 1000 GP).\nGATK_DIR: Path to GATK specific resources (see GATK).\nGRC37_TO_GRC38_CHAIN_FILE (default: \"./assets/hg19ToHg38.over.chain.gz\"): Path to chain file used to liftover the GRCh37 genotypes to GRCh38.","category":"page"},{"location":"combining_genomicc_datasets/#QC-Parameters","page":"Workflow","title":"QC Parameters","text":"","category":"section"},{"location":"combining_genomicc_datasets/","page":"Workflow","title":"Workflow","text":"QC_GENOTYPE_MISSING_RATE (default: 0.02): Maximum missing rate per variant across all individuals. Variants above the threshold are dropped.\nQC_INDIVIDUAL_MISSING_RATE (default: 0.02): Maximum missing rate per individual across genotypes. Individuals above the threshold are dropped.\nQC_HWE_P (default: 1e-10): Used to identify potential technical artifacts and drop variants.\nQC_HWE_K (default: 0.001): Used together with QC_HWE_P\nPCA_APPROX (default: true): Whether PCA is performed via approximation see\nFILTER_HIGH_LOADINGS_VARIANTS (default: false): Whether to drop variants with high PCA loadings. If the loadings plots exhibits a high peak you may want to turn that on.\nANCESTRY_THRESHOLD (default: 0.5): For each individual, the most likely ancestry estimate should be greater than this threshold otherwise the individual is marked as admixed.\nANCESTRY_PROGRAM (default: scope): Program to use for ancestry estimation. Either scope or admixture.","category":"page"},{"location":"combining_genomicc_datasets/#Output-Directories-Parameters","page":"Workflow","title":"Output Directories Parameters","text":"","category":"section"},{"location":"combining_genomicc_datasets/","page":"Workflow","title":"Workflow","text":"PUBLISH_DIR (default: \"results\"): Top level directory where outputs will be output.\nKGP_PUBLISH_DIR (default: \"results/kgp\"): Where 1000 Genome Project data will be output.\nARRAY_GENOTYPES_PUBLISH_DIR (default: \"results/array_genotypes\"): Where data associated with the genotyping arrays will be output.\nWGS_PUBLISH_DIR (default: \"results/wgs\"): Where data associated with the whole-genome sequencing data will be output.\nGATK_PUBLISH_DIR (default: \"results/gatk\"): Where data associated with GATK requirements will be output.\nMERGED_PUBLISH_DIR (default: \"results/merged\"): Where the merged genetic data will be output.","category":"page"},{"location":"combining_genomicc_datasets/#Current-Limitations","page":"Workflow","title":"Current Limitations","text":"","category":"section"},{"location":"combining_genomicc_datasets/","page":"Workflow","title":"Workflow","text":"These are current limitations of the aggregation workflow:","category":"page"},{"location":"combining_genomicc_datasets/","page":"Workflow","title":"Workflow","text":"Only chromosomes 1 to 22 are processed.\nOnly bi-allelic SNPs are used.","category":"page"},{"location":"combining_genomicc_datasets/#Workflow-DAG","page":"Workflow","title":"Workflow DAG","text":"","category":"section"},{"location":"combining_genomicc_datasets/","page":"Workflow","title":"Workflow","text":"<iframe src=\"../assets/combining_datasets_dag.html\" width=\"100%\" height=\"2000px\"></iframe>","category":"page"},{"location":"#GenOMICC-Workflows","page":"Home","title":"GenOMICC Workflows","text":"","category":"section"},{"location":"#Purpose","page":"Home","title":"Purpose","text":"","category":"section"},{"location":"","page":"Home","title":"Home","text":"This repository provides a set of Nextflow workflows that can be used to analyse data generated by the GenOMICC project.","category":"page"},{"location":"#Cohorts","page":"Home","title":"Cohorts","text":"","category":"section"},{"location":"","page":"Home","title":"Home","text":"Even though our main dataset consists of GenoMICC data, for historical reasons (COVID-19 pandemic), we have a total of 4 different cohorts: GenOMICC, MILD, REACT and ISARIC. MILD, REACT and ISARIC were integrated in order to provide non severe \"controls\" for various COVID-19 susceptibility studies.","category":"page"},{"location":"#GenOMICC-(Genetics-Of-Mortality-In-Critical-Care)","page":"Home","title":"GenOMICC (Genetics Of Mortality In Critical Care)","text":"","category":"section"},{"location":"","page":"Home","title":"Home","text":"Link to the study webpage: https://genomicc.org/.","category":"page"},{"location":"#GenOMICC-UK","page":"Home","title":"GenOMICC UK","text":"","category":"section"},{"location":"","page":"Home","title":"Home","text":"Inclusion criteria:","category":"page"},{"location":"","page":"Home","title":"Home","text":"Intensive Care Unit patients\nUK (mostly England and some specific recruitment centers if I understand correctly)","category":"page"},{"location":"","page":"Home","title":"Home","text":"Additional Information:","category":"page"},{"location":"","page":"Home","title":"Home","text":"The study also seems to be recruiting controls by matching in some way according to this page.\nNurses select who they think might be eligible which could produce cryptic bias as well.","category":"page"},{"location":"#GenOMICC-International","page":"Home","title":"GenOMICC International","text":"","category":"section"},{"location":"","page":"Home","title":"Home","text":"This is still being deployed and is not part of our current data.","category":"page"},{"location":"#ISARIC","page":"Home","title":"ISARIC","text":"","category":"section"},{"location":"","page":"Home","title":"Home","text":"Link to the study webpage: https://isaric.org/.","category":"page"},{"location":"","page":"Home","title":"Home","text":"Inclusion criteria:","category":"page"},{"location":"","page":"Home","title":"Home","text":"Our subset of individuals all had Covid-19","category":"page"},{"location":"","page":"Home","title":"Home","text":"Additional Information:","category":"page"},{"location":"","page":"Home","title":"Home","text":"The severity varies and a severity score is provided\nMany biological and contextual covariates exist but are sparse","category":"page"},{"location":"#MILD","page":"Home","title":"MILD","text":"","category":"section"},{"location":"","page":"Home","title":"Home","text":"Link to the study webpage: ?","category":"page"},{"location":"","page":"Home","title":"Home","text":"Inclusion criteria:","category":"page"},{"location":"","page":"Home","title":"Home","text":"Covid-19\nVolunteers (It is not clear in what way?)\nNot hospitalised ","category":"page"},{"location":"#REACT-(Real-time-Assessment-of-Community-Transmission)","page":"Home","title":"REACT (Real-time Assessment of Community Transmission)","text":"","category":"section"},{"location":"","page":"Home","title":"Home","text":"Link to the study webpage https://www.imperial.ac.uk/medicine/research-and-impact/groups/react-study/.","category":"page"},{"location":"","page":"Home","title":"Home","text":"Inclusion criteria:","category":"page"},{"location":"","page":"Home","title":"Home","text":"Covid-19\nIt is unclear if they were hospitalised\nIt is unclear what the selection process is","category":"page"},{"location":"#Available-Workflows","page":"Home","title":"Available Workflows","text":"","category":"section"},{"location":"","page":"Home","title":"Home","text":"These workflows will typically be run sequentially.","category":"page"},{"location":"","page":"Home","title":"Home","text":"Combining GenOMICC Datasets","category":"page"}]
}
