<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Input Data · SequentialGWAS.jl</title><meta name="title" content="Input Data · SequentialGWAS.jl"/><meta property="og:title" content="Input Data · SequentialGWAS.jl"/><meta property="twitter:title" content="Input Data · SequentialGWAS.jl"/><meta name="description" content="Documentation for SequentialGWAS.jl."/><meta property="og:description" content="Documentation for SequentialGWAS.jl."/><meta property="twitter:description" content="Documentation for SequentialGWAS.jl."/><meta property="og:url" content="https://baillielab.github.io/SequentialGWAS.jl/input_data/"/><meta property="twitter:url" content="https://baillielab.github.io/SequentialGWAS.jl/input_data/"/><link rel="canonical" href="https://baillielab.github.io/SequentialGWAS.jl/input_data/"/><script data-outdated-warner src="../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../search_index.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-mocha.css" data-theme-name="catppuccin-mocha"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-macchiato.css" data-theme-name="catppuccin-macchiato"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-frappe.css" data-theme-name="catppuccin-frappe"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-latte.css" data-theme-name="catppuccin-latte"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../">SequentialGWAS.jl</a></span></div><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li><a class="tocitem" href="../">Home</a></li><li><a class="tocitem" href="../cohorts/">Cohorts Description</a></li><li class="is-active"><a class="tocitem" href>Input Data</a><ul class="internal"><li><a class="tocitem" href="#Genetic-Data"><span>Genetic Data</span></a></li><li><a class="tocitem" href="#Covariates-Data"><span>Covariates Data</span></a></li></ul></li><li><a class="tocitem" href="../run/">Run</a></li><li><a class="tocitem" href="../outputs/">Outputs</a></li><li><a class="tocitem" href="../development/">Development</a></li><li><a class="tocitem" href="../mock_data/">Mock Data</a></li><li><a class="tocitem" href="../julia_fns/">Index Of Julia Functions</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>Input Data</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Input Data</a></li></ul></nav><div class="docs-right"><a class="docs-navbar-link" href="https://github.com/baillielab/sequential-gwas" title="View the repository on GitHub"><span class="docs-icon fa-brands"></span><span class="docs-label is-hidden-touch">GitHub</span></a><a class="docs-navbar-link" href="https://github.com/baillielab/sequential-gwas/blob/main/docs/src/input_data.md" title="Edit source on GitHub"><span class="docs-icon fa-solid"></span></a><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><h1 id="Input-Data"><a class="docs-heading-anchor" href="#Input-Data">Input Data</a><a id="Input-Data-1"></a><a class="docs-heading-anchor-permalink" href="#Input-Data" title="Permalink"></a></h1><p>Since GenOMICC is an ongoing project where the data is continuously collected, it came and will continue to arrive in different formats. This page aims at making it clear what are the inputs and outputs to the pipeline.</p><p>All file paths are currently reported with respect to my a015 project but this will be standardized in the future.</p><h2 id="Genetic-Data"><a class="docs-heading-anchor" href="#Genetic-Data">Genetic Data</a><a id="Genetic-Data-1"></a><a class="docs-heading-anchor-permalink" href="#Genetic-Data" title="Permalink"></a></h2><p>Depending on the time period, individuals may have been genotyped, whole genome sequenced or both.</p><h3 id="Genotyping-Arrays"><a class="docs-heading-anchor" href="#Genotyping-Arrays">Genotyping Arrays</a><a id="Genotyping-Arrays-1"></a><a class="docs-heading-anchor-permalink" href="#Genotyping-Arrays" title="Permalink"></a></h3><table><tr><th style="text-align: center">Brief Description</th><th style="text-align: center">Period Begin</th><th style="text-align: center">Period End</th><th style="text-align: center">Genotyping Array</th><th style="text-align: center">Genome Build</th><th style="text-align: center">Directory</th><th style="text-align: center">Genotypes File</th></tr><tr><td style="text-align: center">Prehistoric r8 release</td><td style="text-align: center">04/05/2020</td><td style="text-align: center">30/08/2021</td><td style="text-align: center">GSA-MD-24v3-0_A1</td><td style="text-align: center">GRC37</td><td style="text-align: center">wp5-gwas-r8-under90excl_2021Sep16</td><td style="text-align: center">PLINK<em>190921</em>0906/wp5-gwas-r8-under90excl_2021Sep16.ped</td></tr><tr><td style="text-align: center">Before 2024 microarray</td><td style="text-align: center">20/09/2021</td><td style="text-align: center">06/12/2023</td><td style="text-align: center">GSA-MD-24v3-0_A1</td><td style="text-align: center">GRC37</td><td style="text-align: center">20210920<em>20231206</em>QC_VFinal</td><td style="text-align: center">PLINK<em>040724</em>0954/20210920<em>20231206</em>QC_VFinal.ped</td></tr><tr><td style="text-align: center">Since 2024 microarray</td><td style="text-align: center">04/06/2024</td><td style="text-align: center">10/06/2024</td><td style="text-align: center">GSA-MD-48v4-0_A1</td><td style="text-align: center">GRC38</td><td style="text-align: center">20240604<em>20240610</em>QC_VFinal</td><td style="text-align: center">PLINK<em>040724</em>0114/20240604<em>20240610</em>QC_VFinal.ped</td></tr></table><p>The Illumina manifest files corresponding to each array can be downloaded from <a href="https://emea.support.illumina.com/array/array_kits/infinium-global-screening-array/downloads.html">Illumina&#39;s website</a>. The description of the manifest columns can be found <a href="https://knowledge.illumina.com/microarray/general/microarray-general-reference_material-list/000001565">here</a>. A comparison of both the GSA-48v4 and GSA-24v3 arrays for the GRC38 genome build yields the following table (function <code>SequentialGWAS.OneTimeChecks.array_overlap</code>): </p><table><tr><th style="text-align: center">Description</th><th style="text-align: center">Number of SNPs</th></tr><tr><td style="text-align: center">In GSA-48v4</td><td style="text-align: center">650 321</td></tr><tr><td style="text-align: center">In GSA-24v3</td><td style="text-align: center">654 027</td></tr><tr><td style="text-align: center">Union</td><td style="text-align: center">702 515</td></tr><tr><td style="text-align: center">Intersection</td><td style="text-align: center">601 833</td></tr><tr><td style="text-align: center">Only in GSA-48v4</td><td style="text-align: center">48 488</td></tr><tr><td style="text-align: center">Only in GSA-24v3</td><td style="text-align: center">52 194</td></tr></table><p>The arrays are thus quite similar with around 7% difference between the two and none being a subset of the other. This motivates the strategy to take the intersection of genotyping arrays before merging. Note that in reality the intersection might be smaller due to QC filtering of the input SNPs.</p><p>Furthermore, because the manifest files are too big to be version controlled, variants on the - RefStrand were extracted (function <code>SequentialGWAS.OneTimeChecks.make_snps_to_flip_list</code>) and stored as follows:</p><ul><li>GSA-MD-24v3-0<em>A1 (genome build GRC37) : `assets/GSA-24v3-0</em>A1-minus-strand.txt`</li><li>GSA-MD-48v4-0<em>A1 (genome build GRC38) : `assets/GSA-48v4-0</em>20085471_D2-minus-strand.txt`</li></ul><p>This has been done using the following command:</p><pre><code class="language-bash hljs">julia --project --startup-file=no bin/seq-gwas.jl snps-to-flip --help</code></pre><h3 id="Whole-Genome-Sequencing"><a class="docs-heading-anchor" href="#Whole-Genome-Sequencing">Whole Genome Sequencing</a><a id="Whole-Genome-Sequencing-1"></a><a class="docs-heading-anchor-permalink" href="#Whole-Genome-Sequencing" title="Permalink"></a></h3><p>Currently the data available to me is a subset of ~500 000 SNPs split in files of groups of individuals in <code>/odp-beegfs/a015/linked_data/preqc/gws-genotyped</code>.</p><h3 id="Erola&#39;s-Historical-Data-(This-paragraph-will-be-deleted)"><a class="docs-heading-anchor" href="#Erola&#39;s-Historical-Data-(This-paragraph-will-be-deleted)">Erola&#39;s Historical Data (This paragraph will be deleted)</a><a id="Erola&#39;s-Historical-Data-(This-paragraph-will-be-deleted)-1"></a><a class="docs-heading-anchor-permalink" href="#Erola&#39;s-Historical-Data-(This-paragraph-will-be-deleted)" title="Permalink"></a></h3><p>As currently understood and relative to the data filesystem in <code>/odp-beegfs/a015/postqc/array/</code>:</p><ul><li><code>imputation-output/chr*.info.gz</code> are outputs from TopMed imputation services from <code>erola_scripts/downloadTOPMED.sh</code>. These originate from merged/liftover genotyped files that have been lost.</li><li><code>imputation-output/Genomicc_chr*.txt</code> are probable outputs from <code>erola_scripts/filtervcf.sh</code> based on the above even though there seems to be an intermediate vcf file that was deleted.</li><li>In <code>bgen/</code> are the likely output of <code>erola_scripts/vcftobgen.sh</code>.</li></ul><h2 id="Covariates-Data"><a class="docs-heading-anchor" href="#Covariates-Data">Covariates Data</a><a id="Covariates-Data-1"></a><a class="docs-heading-anchor-permalink" href="#Covariates-Data" title="Permalink"></a></h2><p>The covariate file is a CSV file in <code>/odp-beegfs/a015/linked_data/postqc/array/a015_covariates.csv</code> with the following columns:</p><ul><li>ODAP_ID: Unique invidual identifier (same as the IID/FID in the genetic data).</li><li>COHORTS: List of all cohorts the individual belongs to.</li><li>AGE<em>AT</em>RECRUITMENT: Age of the individual at recruitment. If an individual was recruited in two cohorts, typically (GenOMICC and ISARIC), GenOMICC takes precedence.</li><li>REPORTED_SEX: Sex reported by an individual.</li><li>PRIMARY_DIAGNOSIS: Only for GenOMICC patients and obtained via REDCap.</li><li>DIAGNOSES: List of diagnoses inferred from REDCap</li><li>ISARIC<em>SEVERITY</em>SCORE: Only for ISARIC individuals, severity of infection.</li><li>GENETIC<em>SAMPLE</em>TYPE: Blood or Saliva.</li><li>CALL_RATE: Genotyping call rate (saliva samples are believed to have lower quality).</li><li>GENETIC<em>MEASUREMENT</em>TECHNOLOGY: At this time either of (GSA-MD-24v3-0<em>A1, GSA-MD-48v4-0</em>A1, WGS).</li><li>DATE<em>AT</em>RECRUITMENT: Time marker of individual recruitement.</li></ul></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../cohorts/">« Cohorts Description</a><a class="docs-footer-nextpage" href="../run/">Run »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="auto">Automatic (OS)</option><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option><option value="catppuccin-latte">catppuccin-latte</option><option value="catppuccin-frappe">catppuccin-frappe</option><option value="catppuccin-macchiato">catppuccin-macchiato</option><option value="catppuccin-mocha">catppuccin-mocha</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.8.1 on <span class="colophon-date" title="Saturday 15 February 2025 17:09">Saturday 15 February 2025</span>. Using Julia version 1.11.3.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
