<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Merging the GenOMICC and UK Biobank Cohorts · SequentialGWAS.jl</title><meta name="title" content="Merging the GenOMICC and UK Biobank Cohorts · SequentialGWAS.jl"/><meta property="og:title" content="Merging the GenOMICC and UK Biobank Cohorts · SequentialGWAS.jl"/><meta property="twitter:title" content="Merging the GenOMICC and UK Biobank Cohorts · SequentialGWAS.jl"/><meta name="description" content="Documentation for SequentialGWAS.jl."/><meta property="og:description" content="Documentation for SequentialGWAS.jl."/><meta property="twitter:description" content="Documentation for SequentialGWAS.jl."/><meta property="og:url" content="https://baillielab.github.io/SequentialGWAS.jl/ukb_merge/"/><meta property="twitter:url" content="https://baillielab.github.io/SequentialGWAS.jl/ukb_merge/"/><link rel="canonical" href="https://baillielab.github.io/SequentialGWAS.jl/ukb_merge/"/><script data-outdated-warner src="../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../search_index.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-mocha.css" data-theme-name="catppuccin-mocha"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-macchiato.css" data-theme-name="catppuccin-macchiato"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-frappe.css" data-theme-name="catppuccin-frappe"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-latte.css" data-theme-name="catppuccin-latte"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../">SequentialGWAS.jl</a></span></div><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li><a class="tocitem" href="../">Home</a></li><li><span class="tocitem">Workflows</span><ul><li><a class="tocitem" href="../prerequisites/">Prerequisites</a></li><li><a class="tocitem" href="../combining_datasets/">Combining Datasets</a></li><li><a class="tocitem" href="../genotypes_imputation/">Genotypes Imputation</a></li><li class="is-active"><a class="tocitem" href>Merging the GenOMICC and UK Biobank Cohorts</a><ul class="internal"><li><a class="tocitem" href="#Inputs"><span>Inputs</span></a></li><li><a class="tocitem" href="#Workflow-Parameters"><span>Workflow Parameters</span></a></li><li><a class="tocitem" href="#Running-The-Workflow"><span>Running The Workflow</span></a></li><li><a class="tocitem" href="#Outputs"><span>Outputs</span></a></li></ul></li><li><a class="tocitem" href="../gwas/">GWAS</a></li></ul></li><li><span class="tocitem">For Developpers</span><ul><li><a class="tocitem" href="../development/">Development</a></li><li><a class="tocitem" href="../mock_data/">Mock Data</a></li><li><a class="tocitem" href="../julia_fns/">Index Of Julia Functions</a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">Workflows</a></li><li class="is-active"><a href>Merging the GenOMICC and UK Biobank Cohorts</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Merging the GenOMICC and UK Biobank Cohorts</a></li></ul></nav><div class="docs-right"><a class="docs-navbar-link" href="https://github.com/baillielab/sequential-gwas" title="View the repository on GitHub"><span class="docs-icon fa-brands"></span><span class="docs-label is-hidden-touch">GitHub</span></a><a class="docs-navbar-link" href="https://github.com/baillielab/sequential-gwas/blob/main/docs/src/ukb_merge.md" title="Edit source on GitHub"><span class="docs-icon fa-solid"></span></a><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><h1 id="Merging-the-GenOMICC-and-UK-Biobank-Cohorts"><a class="docs-heading-anchor" href="#Merging-the-GenOMICC-and-UK-Biobank-Cohorts">Merging the GenOMICC and UK Biobank Cohorts</a><a id="Merging-the-GenOMICC-and-UK-Biobank-Cohorts-1"></a><a class="docs-heading-anchor-permalink" href="#Merging-the-GenOMICC-and-UK-Biobank-Cohorts" title="Permalink"></a></h1><p>This workflow yields combined:</p><ul><li>genotypes in PLINK bed format: obtained by subsampling the TOPMed UKB imputed genotypes to keep the GenOMICC variants.</li><li>imputed genotypes in PLINK pgen format: ontained by merging UKB and GenOMICC TOPMed UKB imputed genotypes.</li><li>covariates containing the current age and sex of individuals</li><li>ancestry estimates: via the 1000 Genome Project</li></ul><h2 id="Inputs"><a class="docs-heading-anchor" href="#Inputs">Inputs</a><a id="Inputs-1"></a><a class="docs-heading-anchor-permalink" href="#Inputs" title="Permalink"></a></h2><p>This section describes the main inputs to the workflow, see further down for the associated Nextflow parameters.</p><h3 id="GenOMICC-Data"><a class="docs-heading-anchor" href="#GenOMICC-Data">GenOMICC Data</a><a id="GenOMICC-Data-1"></a><a class="docs-heading-anchor-permalink" href="#GenOMICC-Data" title="Permalink"></a></h3><p>For this workflow, genotypes, imputed genotypes are required. They will be the respective outputs of the <a href="../combining_datasets/#Combining-Datasets">Combining Datasets</a> and <a href="../genotypes_imputation/#Genotypes-Imputation">Genotypes Imputation</a> workflows.</p><p>You will also need two covariates files:</p><ul><li>The <code>COVARIATES</code> file contains the usual covariates information including phenotypes and is typically provided by Dominique on a per project basis. Unfortunately, at this point, the specifications for this file have not been clearly defined. The columns of the file I have received are: <code>genotype_file_id</code>, <code>age_years</code>, <code>sex</code>, <code>case_or_control</code>, <code>cohort</code>, <code>severe_cohort_primary_diagnosis</code>, <code>isaric_cohort_max_severity_score</code>.</li><li>An optional <code>INFERRED_COVARIATES</code> contains covariates information that is inferred from the genetic datasets during the <a href="../combining_datasets/#Combining-Datasets">Combining Datasets</a> workflow. This is the case for ancestry estimates for instance.</li></ul><h3 id="1000-Genome-Project"><a class="docs-heading-anchor" href="#1000-Genome-Project">1000 Genome Project</a><a id="1000-Genome-Project-1"></a><a class="docs-heading-anchor-permalink" href="#1000-Genome-Project" title="Permalink"></a></h3><p>You will also need the 1000 Ggenome Project genotypes in plink format, this is typically output by the <a href="../combining_datasets/#Combining-Datasets">Combining Datasets</a> workflow, but can also be obtained by running the yet undocumented Nextflow <code>KGP</code> workflow in this repository.</p><h3 id="UK-Biobank-data"><a class="docs-heading-anchor" href="#UK-Biobank-data">UK Biobank data</a><a id="UK-Biobank-data-1"></a><a class="docs-heading-anchor-permalink" href="#UK-Biobank-data" title="Permalink"></a></h3><p>The UK-Biobank data is on the RAP, there is nothing you need to do but to setup your account and follow the instructions in the <a href="../dnanexus/#Working-with-DNANexus">Working with DNANexus</a> section.</p><h2 id="Workflow-Parameters"><a class="docs-heading-anchor" href="#Workflow-Parameters">Workflow Parameters</a><a id="Workflow-Parameters-1"></a><a class="docs-heading-anchor-permalink" href="#Workflow-Parameters" title="Permalink"></a></h2><p>In <code>wdl/ukb_merge/inputs.json</code>, find and replace all instances of the project-ID with your own.</p><h2 id="Running-The-Workflow"><a class="docs-heading-anchor" href="#Running-The-Workflow">Running The Workflow</a><a id="Running-The-Workflow-1"></a><a class="docs-heading-anchor-permalink" href="#Running-The-Workflow" title="Permalink"></a></h2><p>First you need to compile the WDL workflow and upload it to the RAP, this can be done by the following:</p><pre><code class="language-bash hljs">export DX_COMPILER_PATH=/Users/olabayle/dxCompiler/dxCompiler-2.13.0.jar
export PROJECT_ID=project-J0pkqyQJpYQ133JG1p2J1qzv
java -jar $DX_COMPILER_PATH compile wdl/ukb_merge/workflow.wdl -f -project $PROJECT_ID -folder /workflows -inputs wdl/ukb_merge/inputs.json</code></pre><p>where the <code>DX_COMPILER_PATH</code> and <code>PROJECT_ID</code> have to be set appropriately. The compiler might output some warnings like <code>missing input for non-optional parameter</code> but you can ignore these.</p><p>Then, you can run the workflow with the following command</p><pre><code class="language-bash hljs">dx run /workflows/merge_ukb_and_genomicc -f wdl/ukb_merge/inputs.dx.json --destination /ukb_merge_outputs/</code></pre><h2 id="Outputs"><a class="docs-heading-anchor" href="#Outputs">Outputs</a><a id="Outputs-1"></a><a class="docs-heading-anchor-permalink" href="#Outputs" title="Permalink"></a></h2></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../genotypes_imputation/">« Genotypes Imputation</a><a class="docs-footer-nextpage" href="../gwas/">GWAS »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="auto">Automatic (OS)</option><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option><option value="catppuccin-latte">catppuccin-latte</option><option value="catppuccin-frappe">catppuccin-frappe</option><option value="catppuccin-macchiato">catppuccin-macchiato</option><option value="catppuccin-mocha">catppuccin-mocha</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.13.0 on <span class="colophon-date" title="Wednesday 25 June 2025 16:00">Wednesday 25 June 2025</span>. Using Julia version 1.11.5.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
