<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Index Of Julia Functions · SequentialGWAS.jl</title><meta name="title" content="Index Of Julia Functions · SequentialGWAS.jl"/><meta property="og:title" content="Index Of Julia Functions · SequentialGWAS.jl"/><meta property="twitter:title" content="Index Of Julia Functions · SequentialGWAS.jl"/><meta name="description" content="Documentation for SequentialGWAS.jl."/><meta property="og:description" content="Documentation for SequentialGWAS.jl."/><meta property="twitter:description" content="Documentation for SequentialGWAS.jl."/><meta property="og:url" content="https://baillielab.github.io/SequentialGWAS.jl/julia_fns/"/><meta property="twitter:url" content="https://baillielab.github.io/SequentialGWAS.jl/julia_fns/"/><link rel="canonical" href="https://baillielab.github.io/SequentialGWAS.jl/julia_fns/"/><script data-outdated-warner src="../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../search_index.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-mocha.css" data-theme-name="catppuccin-mocha"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-macchiato.css" data-theme-name="catppuccin-macchiato"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-frappe.css" data-theme-name="catppuccin-frappe"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-latte.css" data-theme-name="catppuccin-latte"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../">SequentialGWAS.jl</a></span></div><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li><a class="tocitem" href="../">Home</a></li><li><a class="tocitem" href="../cohorts/">Cohorts Description</a></li><li><a class="tocitem" href="../input_data/">Input Data</a></li><li><a class="tocitem" href="../run/">Run</a></li><li><a class="tocitem" href="../outputs/">Outputs</a></li><li><a class="tocitem" href="../development/">Development</a></li><li><a class="tocitem" href="../mock_data/">Mock Data</a></li><li class="is-active"><a class="tocitem" href>Index Of Julia Functions</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>Index Of Julia Functions</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Index Of Julia Functions</a></li></ul></nav><div class="docs-right"><a class="docs-navbar-link" href="https://github.com/baillielab/sequential-gwas" title="View the repository on GitHub"><span class="docs-icon fa-brands"></span><span class="docs-label is-hidden-touch">GitHub</span></a><a class="docs-navbar-link" href="https://github.com/baillielab/sequential-gwas/blob/main/docs/src/julia_fns.md" title="Edit source on GitHub"><span class="docs-icon fa-solid"></span></a><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><h1 id="Index-of-Julia-Functions"><a class="docs-heading-anchor" href="#Index-of-Julia-Functions">Index of Julia Functions</a><a id="Index-of-Julia-Functions-1"></a><a class="docs-heading-anchor-permalink" href="#Index-of-Julia-Functions" title="Permalink"></a></h1><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SequentialGWAS.get_action-Tuple{Any}" href="#SequentialGWAS.get_action-Tuple{Any}"><code>SequentialGWAS.get_action</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><p>Compares the variant to the 1000 GP information and returns an action to take together with a reason.</p><p>Potential actions are</p><ol><li>DROP</li><li>KEEP</li><li>FLIP</li></ol><p>Some particularly unexpected ACTION (REASON) are:</p><ul><li>&quot;KEEP (REVERSE-REF-ALT)&quot;</li><li>&quot;FLIP (COMPLEMENT-NOT-MATCHING-KGP)&quot;</li></ul><p>Because they mean the minor/major alleles are reversed in our dataset as compared to the reference KGP.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/baillielab/sequential-gwas/blob/dec7095abd4d49a86ff91c800c4926d26d8e36fd/src/qc_from_kgp.jl#L4-L18">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SequentialGWAS.kgp_unrelated_individuals-Tuple{Any}" href="#SequentialGWAS.kgp_unrelated_individuals-Tuple{Any}"><code>SequentialGWAS.kgp_unrelated_individuals</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><p>Only keeps the first individual within each family and writes them to <code>outfile</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/baillielab/sequential-gwas/blob/dec7095abd4d49a86ff91c800c4926d26d8e36fd/src/relatedness.jl#L1-L4">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SequentialGWAS.read_bim-Tuple{Any}" href="#SequentialGWAS.read_bim-Tuple{Any}"><code>SequentialGWAS.read_bim</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">read_bim(file)</code></pre><p>Columns Description from: https://www.cog-genomics.org/plink/1.9/formats#bim</p><ul><li>Chromosome code (either an integer, or &#39;X&#39;/&#39;Y&#39;/&#39;XY&#39;/&#39;MT&#39;; &#39;0&#39; indicates unknown) or name</li><li>Variant identifier</li><li>Position in morgans or centimorgans (safe to use dummy value of &#39;0&#39;)</li><li>Base-pair coordinate (1-based; limited to 231-2)</li><li>Allele 1 (corresponding to clear bits in .bed; usually minor)</li><li>Allele 2 (corresponding to set bits in .bed; usually major)</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/baillielab/sequential-gwas/blob/dec7095abd4d49a86ff91c800c4926d26d8e36fd/src/read_write.jl#L2-L12">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SequentialGWAS.read_fam-Tuple{Any}" href="#SequentialGWAS.read_fam-Tuple{Any}"><code>SequentialGWAS.read_fam</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">read_fam(file)</code></pre><p>Columns Description from: https://www.cog-genomics.org/plink/1.9/formats#fam</p><ul><li>Family ID (&#39;FID&#39;)</li><li>Within-family ID (&#39;IID&#39;; cannot be &#39;0&#39;)</li><li>Within-family ID of father (&#39;0&#39; if father isn&#39;t in dataset)</li><li>Within-family ID of mother (&#39;0&#39; if mother isn&#39;t in dataset)</li><li>Sex code (&#39;1&#39; = male, &#39;2&#39; = female, &#39;0&#39; = unknown)</li><li>Phenotype value (&#39;1&#39; = control, &#39;2&#39; = case, &#39;-9&#39;/&#39;0&#39;/non-numeric = missing data if case/control)</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/baillielab/sequential-gwas/blob/dec7095abd4d49a86ff91c800c4926d26d8e36fd/src/read_write.jl#L20-L30">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SequentialGWAS.read_map-Tuple{Any}" href="#SequentialGWAS.read_map-Tuple{Any}"><code>SequentialGWAS.read_map</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">read_map(file)</code></pre><p>Columns Description from: https://www.cog-genomics.org/plink/1.9/formats</p><ul><li>Chromosome code. PLINK 1.9 also permits contig names here, but most older programs do not.</li><li>Variant identifier</li><li>Position in morgans or centimorgans (optional; also safe to use dummy value of &#39;0&#39;)</li><li>Base-pair coordinate</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/baillielab/sequential-gwas/blob/dec7095abd4d49a86ff91c800c4926d26d8e36fd/src/read_write.jl#L45-L54">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SequentialGWAS.report_qc_effect-Tuple{Any, Any}" href="#SequentialGWAS.report_qc_effect-Tuple{Any, Any}"><code>SequentialGWAS.report_qc_effect</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">report_qc_effect(input_prefix, output_prefix)</code></pre><p>Report the effect of the QC filtering process on SNPs and samples.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/baillielab/sequential-gwas/blob/dec7095abd4d49a86ff91c800c4926d26d8e36fd/src/report.jl#L16-L20">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SequentialGWAS.write_map-Tuple{Any, Any}" href="#SequentialGWAS.write_map-Tuple{Any, Any}"><code>SequentialGWAS.write_map</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">write_map(file_prefix, array)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/baillielab/sequential-gwas/blob/dec7095abd4d49a86ff91c800c4926d26d8e36fd/src/read_write.jl#L57-L59">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SequentialGWAS.write_release_samples_to_drop-Tuple{Any, Any}" href="#SequentialGWAS.write_release_samples_to_drop-Tuple{Any, Any}"><code>SequentialGWAS.write_release_samples_to_drop</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><p>We drop duplicate individuals according to the following priority:</p><p>WGS &gt; More Recent Array &gt; Older Array</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/baillielab/sequential-gwas/blob/dec7095abd4d49a86ff91c800c4926d26d8e36fd/src/qc_from_kgp.jl#L161-L165">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SequentialGWAS.OneTimeChecks.array_overlap-Tuple{Any, Any}" href="#SequentialGWAS.OneTimeChecks.array_overlap-Tuple{Any, Any}"><code>SequentialGWAS.OneTimeChecks.array_overlap</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">array_overlap(manifest_file1, manifest_file2)</code></pre><p>Takes two manifest files from Illumina genotyping arrays and returns the intersection of the SNP names.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/baillielab/sequential-gwas/blob/dec7095abd4d49a86ff91c800c4926d26d8e36fd/src/one_time_checks.jl#L56-L60">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SequentialGWAS.OneTimeChecks.identify_snps_to_flip-Tuple{Any}" href="#SequentialGWAS.OneTimeChecks.identify_snps_to_flip-Tuple{Any}"><code>SequentialGWAS.OneTimeChecks.identify_snps_to_flip</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">identify_snps_to_flip(manifest_file)</code></pre><p>According to <a href="https://knowledge.illumina.com/microarray/general/microarray-general-reference_material-list/000001565">this link</a>,  the RefStrand column in the manifest file corresponds to the standard designation for all eukaryotic organisms used by HapMap and 1000 Genomes Project. Variants with RefStrand equal to <code>-</code> need to be flipped to the + strand.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/baillielab/sequential-gwas/blob/dec7095abd4d49a86ff91c800c4926d26d8e36fd/src/one_time_checks.jl#L29-L35">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SequentialGWAS.OneTimeChecks.make_snps_to_flip_list-Tuple{Any, Any}" href="#SequentialGWAS.OneTimeChecks.make_snps_to_flip_list-Tuple{Any, Any}"><code>SequentialGWAS.OneTimeChecks.make_snps_to_flip_list</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">make_snps_to_flip_list(output, manifest_file)</code></pre><p>Takes a manifest file from an Illumina genotyping array and writes a list of SNPs  that are on the - strand and need to be flipped to the + strand.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/baillielab/sequential-gwas/blob/dec7095abd4d49a86ff91c800c4926d26d8e36fd/src/one_time_checks.jl#L41-L46">source</a></section></article></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../mock_data/">« Mock Data</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="auto">Automatic (OS)</option><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option><option value="catppuccin-latte">catppuccin-latte</option><option value="catppuccin-frappe">catppuccin-frappe</option><option value="catppuccin-macchiato">catppuccin-macchiato</option><option value="catppuccin-mocha">catppuccin-mocha</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.8.1 on <span class="colophon-date" title="Thursday 6 March 2025 17:33">Thursday 6 March 2025</span>. Using Julia version 1.11.3.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
