<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Run · SequentialGWAS.jl</title><meta name="title" content="Run · SequentialGWAS.jl"/><meta property="og:title" content="Run · SequentialGWAS.jl"/><meta property="twitter:title" content="Run · SequentialGWAS.jl"/><meta name="description" content="Documentation for SequentialGWAS.jl."/><meta property="og:description" content="Documentation for SequentialGWAS.jl."/><meta property="twitter:description" content="Documentation for SequentialGWAS.jl."/><meta property="og:url" content="https://baillielab.github.io/SequentialGWAS.jl/run/"/><meta property="twitter:url" content="https://baillielab.github.io/SequentialGWAS.jl/run/"/><link rel="canonical" href="https://baillielab.github.io/SequentialGWAS.jl/run/"/><script data-outdated-warner src="../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../search_index.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-mocha.css" data-theme-name="catppuccin-mocha"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-macchiato.css" data-theme-name="catppuccin-macchiato"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-frappe.css" data-theme-name="catppuccin-frappe"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-latte.css" data-theme-name="catppuccin-latte"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../">SequentialGWAS.jl</a></span></div><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li><a class="tocitem" href="../">Home</a></li><li><a class="tocitem" href="../cohorts/">Cohorts Description</a></li><li><a class="tocitem" href="../input_data/">Input Data</a></li><li class="is-active"><a class="tocitem" href>Run</a><ul class="internal"><li><a class="tocitem" href="#Dependencies"><span>Dependencies</span></a></li><li><a class="tocitem" href="#Running-The-Workflow-2"><span>Running The Workflow</span></a></li><li><a class="tocitem" href="#Pipeline-parameters"><span>Pipeline parameters</span></a></li></ul></li><li><a class="tocitem" href="../outputs/">Outputs</a></li><li><a class="tocitem" href="../development/">Development</a></li><li><a class="tocitem" href="../mock_data/">Mock Data</a></li><li><a class="tocitem" href="../julia_fns/">Index Of Julia Functions</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>Run</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Run</a></li></ul></nav><div class="docs-right"><a class="docs-navbar-link" href="https://github.com/baillielab/sequential-gwas" title="View the repository on GitHub"><span class="docs-icon fa-brands"></span><span class="docs-label is-hidden-touch">GitHub</span></a><a class="docs-navbar-link" href="https://github.com/baillielab/sequential-gwas/blob/main/docs/src/run.md" title="Edit source on GitHub"><span class="docs-icon fa-solid"></span></a><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><h1 id="Running-The-Workflow"><a class="docs-heading-anchor" href="#Running-The-Workflow">Running The Workflow</a><a id="Running-The-Workflow-1"></a><a class="docs-heading-anchor-permalink" href="#Running-The-Workflow" title="Permalink"></a></h1><h2 id="Dependencies"><a class="docs-heading-anchor" href="#Dependencies">Dependencies</a><a id="Dependencies-1"></a><a class="docs-heading-anchor-permalink" href="#Dependencies" title="Permalink"></a></h2><h3 id="Platform:-ODAP"><a class="docs-heading-anchor" href="#Platform:-ODAP">Platform: ODAP</a><a id="Platform:-ODAP-1"></a><a class="docs-heading-anchor-permalink" href="#Platform:-ODAP" title="Permalink"></a></h3><p>Since the data is sensitive, it is meant to run on the <a href="https://odap.ac.uk/">ODAP</a> platform (at least for now). To get access, you will need to contact the relevant person, at the moment <a href="mailto:dominique.mccormick@ed.ac.uk">dominique.mccormick@ed.ac.uk</a>. More information on accessing ODAP can be found <a href="https://git.ecdf.ed.ac.uk/odap-users-guide/odap-users-guide">here</a>.</p><h3 id="Data"><a class="docs-heading-anchor" href="#Data">Data</a><a id="Data-1"></a><a class="docs-heading-anchor-permalink" href="#Data" title="Permalink"></a></h3><p>Once access to ODAP is given you will also need access to the <a href="../input_data/#Input-Data">Input Data</a>. This is also handled by Dominique at the moment.</p><h3 id="Software"><a class="docs-heading-anchor" href="#Software">Software</a><a id="Software-1"></a><a class="docs-heading-anchor-permalink" href="#Software" title="Permalink"></a></h3><p>In order to run the workflows in this repository only 2 software need to be installed, and they should already be present on ODAP:</p><ul><li>Nextflow 24.10.3 (<a href="https://git.ecdf.ed.ac.uk/odap-users-guide/odap-users-guide/-/wikis/nexflow">see how to setup</a>)</li><li>Singularity 3.9.4 (Should be ready to use)</li></ul><h2 id="Running-The-Workflow-2"><a class="docs-heading-anchor" href="#Running-The-Workflow-2">Running The Workflow</a><a class="docs-heading-anchor-permalink" href="#Running-The-Workflow-2" title="Permalink"></a></h2><p>If the previous steps have been completed successfully you can run:</p><pre><code class="language-bash hljs">nextflow run main.nf -profile odap -resume</code></pre><h2 id="Pipeline-parameters"><a class="docs-heading-anchor" href="#Pipeline-parameters">Pipeline parameters</a><a id="Pipeline-parameters-1"></a><a class="docs-heading-anchor-permalink" href="#Pipeline-parameters" title="Permalink"></a></h2><p>We now list all the pipeline parameters. In principle they don&#39;t need to be changed if the conventions in this documentation have been respected and are up to date. Otherwise, please feel free to run an issue.</p><h3 id="Output-Directories-Parameters"><a class="docs-heading-anchor" href="#Output-Directories-Parameters">Output Directories Parameters</a><a id="Output-Directories-Parameters-1"></a><a class="docs-heading-anchor-permalink" href="#Output-Directories-Parameters" title="Permalink"></a></h3><ul><li>PUBLISH_DIR (default: &quot;results&quot;): Top level directory where outputs will be output.</li><li>KGP<em>PUBLISH</em>DIR (default: &quot;{params.PUBLISH_DIR}/kgp&quot;): Where 1000 Genome Project data will be output.</li><li>ARRAY<em>GENOTYPES</em>PUBLISH<em>DIR (default: &quot;$(Expr(:incomplete, Base.Meta.ParseError(&quot;ParseError:\n# Error @ none:1:16\n{params.PUBLISH\n#              └ ── Expected `}`&quot;, Base.JuliaSyntax.ParseError(Base.JuliaSyntax.SourceFile(&quot;{params.PUBLISH&quot;, 0, &quot;none&quot;, 1, [1, 16]), Base.JuliaSyntax.Diagnostic[Base.JuliaSyntax.Diagnostic(16, 15, :error, &quot;Expected `}`&quot;)], :other))))</em>DIR}/array_genotypes&quot;): Where data associated with the genotyping arrays will be output.</li><li>WGS<em>PUBLISH</em>DIR (default: &quot;{params.PUBLISH_DIR}/wgs&quot;): Where data associated with the whole-genome sequencing data will be output.</li><li>GATK<em>PUBLISH</em>DIR (default: &quot;{params.PUBLISH_DIR}/gatk&quot;): Where data associated with GATK requirements will be output.</li><li>MERGED<em>PUBLISH</em>DIR (default: &quot;{params.PUBLISH_DIR}/merged&quot;): Where the merged genetic data will be output.</li></ul><h3 id="External-Inputs-Parameters"><a class="docs-heading-anchor" href="#External-Inputs-Parameters">External Inputs Parameters</a><a id="External-Inputs-Parameters-1"></a><a class="docs-heading-anchor-permalink" href="#External-Inputs-Parameters" title="Permalink"></a></h3><ul><li>RESOURCES_DIR (default: {projectDir}/assets/resources&quot;): Path to all external resources.</li><li>KGP<em>DIR (default: &quot;$(Expr(:incomplete, Base.Meta.ParseError(&quot;ParseError:\n# Error @ none:1:18\n{params.RESOURCES\n#                └ ── Expected `}`&quot;, Base.JuliaSyntax.ParseError(Base.JuliaSyntax.SourceFile(&quot;{params.RESOURCES&quot;, 0, &quot;none&quot;, 1, [1, 18]), Base.JuliaSyntax.Diagnostic[Base.JuliaSyntax.Diagnostic(18, 17, :error, &quot;Expected `}`&quot;)], :other))))</em>DIR}/kgp&quot;): Path to the 1000 Genome Project specific resources.</li><li>GATK<em>DIR (default: &quot;$(Expr(:incomplete, Base.Meta.ParseError(&quot;ParseError:\n# Error @ none:1:18\n{params.RESOURCES\n#                └ ── Expected `}`&quot;, Base.JuliaSyntax.ParseError(Base.JuliaSyntax.SourceFile(&quot;{params.RESOURCES&quot;, 0, &quot;none&quot;, 1, [1, 18]), Base.JuliaSyntax.Diagnostic[Base.JuliaSyntax.Diagnostic(18, 17, :error, &quot;Expected `}`&quot;)], :other))))</em>DIR}/gatk&quot;): Path to GATK specific resources.</li><li>GRC37<em>TO</em>GRC38<em>CHAIN</em>FILE (default: &quot;{projectDir}/assets/hg19ToHg38.over.chain.gz&quot;): Path to chain file used to liftover the GRCh37 genotypes to GRCh38.</li><li>VARIANTS<em>TO</em>FLIP<em>GRC38 (default: &quot;{projectDir}/assets/GSA-48v4-0</em>20085471_D2-minus-strand.txt&quot;): THis is not used at the moment and probably needs to be dropped.</li><li>VARIANTS<em>TO</em>FLIP<em>GRC37 (default: &quot;{projectDir}/assets/GSA-24v3-0</em>A1-minus-strand.txt&quot;): same as above.</li></ul><h3 id="Basic-QC-Parameters"><a class="docs-heading-anchor" href="#Basic-QC-Parameters">Basic QC Parameters</a><a id="Basic-QC-Parameters-1"></a><a class="docs-heading-anchor-permalink" href="#Basic-QC-Parameters" title="Permalink"></a></h3><ul><li>QC<em>GENOTYPE</em>MISSING_RATE (default: 0.1): Maximum missing rate per variant across all individuals. Variants above the threshold are dropped.</li><li>QC<em>INDIVIDUAL</em>MISSING_RATE (default: 0.1): Maximum missing rate per individual across genotypes. Individuals above the threshold are dropped.</li><li>QC_HWE (default: 1e-50): Used to identify potential technical artifacts and drop variants.</li></ul><h3 id="GWAS-Parameters"><a class="docs-heading-anchor" href="#GWAS-Parameters">GWAS Parameters</a><a id="GWAS-Parameters-1"></a><a class="docs-heading-anchor-permalink" href="#GWAS-Parameters" title="Permalink"></a></h3><ul><li>QC_MAF (default: 0.01): Minor Allele Frequency required to enter the GWAS pipeline.</li><li>QC_MAC (default: 100): Minor Allele Count for REGENIE execution.</li></ul></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../input_data/">« Input Data</a><a class="docs-footer-nextpage" href="../outputs/">Outputs »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="auto">Automatic (OS)</option><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option><option value="catppuccin-latte">catppuccin-latte</option><option value="catppuccin-frappe">catppuccin-frappe</option><option value="catppuccin-macchiato">catppuccin-macchiato</option><option value="catppuccin-mocha">catppuccin-mocha</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.8.1 on <span class="colophon-date" title="Monday 17 February 2025 16:57">Monday 17 February 2025</span>. Using Julia version 1.11.3.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
